### PR TITLE
recognize jshell files

### PR DESCRIPTION
**Pull request by Konfekt**
Maybe @zzzyxwvut could have a look as well

### CODE REVIEW

**Comment by zzzyxwvut** [comment]
These files are scripts for a REPL tool (see [JEP 222](https://openjdk.org/jeps/222)).   
JShell offers `fc`-like (from POSIX) support with its own  
`/edit` command, and after

```bashrc
export JSHELLEDITOR=vim
```

has been set, you can launch Vim from it on demand to juggle  
history items etc.

For reference, [its scripts guide](https://docs.oracle.com/en/java/javase/23/jshell/scripts.html) and bundled scripts: [1](https://raw.githubusercontent.com/openjdk/jdk/refs/heads/master/src/jdk.jshell/share/classes/jdk/jshell/tool/resources/DEFAULT.jsh), [2](https://raw.githubusercontent.com/openjdk/jdk/refs/heads/master/src/jdk.jshell/share/classes/jdk/jshell/tool/resources/PREVIEW_DEFAULT.jsh),  
[3](https://raw.githubusercontent.com/openjdk/jdk/refs/heads/master/src/jdk.jshell/share/classes/jdk/jshell/tool/resources/PRINTING.jsh), [4](https://raw.githubusercontent.com/openjdk/jdk/refs/heads/master/src/jdk.jshell/share/classes/jdk/jshell/tool/resources/TOOLING.jsh).

(Note that setting `g:java_highlight_functions` would prove  
ineffective for these files unless method declarations are  
indented.  Sadly, it can't be changed without breaking the  
`indent` style variant.)

**Comment by chrisbra** [comment]
So do you approve this change @zzzyxwvut ?

File: runtime/ftplugin/java.vim [hunk 1]
```diff
@@ -78,6 +78,10 @@ setlocal comments& comments^=:///,sO:*\ -,mO:*\ \ ,exO:*/
 
 setlocal commentstring=//\ %s
 
+if expand('%:e') ==# 'jsh'
+  match Comment /^#!.*/
```
Comment (zzzyxwvut) [inline runtime/ftplugin/java.vim]: Please can you limit the match to some arbitrary top lines,  
e.g. `\%<32l^#!.*`, and make the indentation conform to the  
modeline?

**Comment by zzzyxwvut** [comment]
I have no clue about the shebang support and JShell's manual  
page doesn't say much either.  I can't get it to work with:

```shell
vim hello.jsh
```

```java
#!/usr/bin/jshell

System.out.println("hello world");
```

```shell
chmod +x hello.jsh
./hello.jsh
```

```
Error:
illegal character: '#'
#!/usr/bin/jshell
^
Error:
illegal start of expression
#!/usr/bin/jshell
  ^
hello world
|  Welcome to JShell -- Version 21.0.5
|  For an introduction type: /help intro

jshell> /exit
|  Goodbye
```

@Konfekt, can you bump the date of `ftplugin/java.vim`?

Also, the `#` is marked as `javaError2` in the above example.  
I guess a `syntax` file extending Java is in order.

**Comment by Konfekt** [comment]
Actually it must be a triple slash instead of a shebang, see [these comments](https://unix.stackexchange.com/a/162535) and  [this answer](https://stackoverflow.com/a/44925456/3528522) for executable jshell scripts. So no shebang detection needed as

```vim
if expand('%:e') ==# 'jsh'
    " See https://stackoverflow.com/a/44925456/3528522
    match Comment /\%<2l^\/\/\/.*/
endif
```

is redundant, since `//` is already a comment marker

**Comment by Konfekt** [comment]
One could think about `runtime/autoload/dist/script.vim` detecting triple slash "shebangs", but that's another can of worms

**Comment by zzzyxwvut** [comment]
Alright, so there is no ‘shebang’ support, users ought to  
launch `jshell` and from its prompt source JShell scripts with  
`/open path/to/foo.jsh` (like `. path/to/foo` in a POSIX shell)  
or pass such arguments to the utility, e.g. `jshell foo.jsh`.

What the linked discussions suggest belongs to what a shell  
and its host OS are capable of when arbitrary file execution  
is attempted.  The linked answer shows this kind of syntax  
(note the two leading slashes in the ‘shebang’ line):

```java
//usr/bin/env jshell --execution local "$0" "$@"; exit $?

System.out.println("hello world")
/exit
```

(Remember that the triple `///` ‘shebang’ line will be matched  
as `javaMarkdownComment` unless either `g:java_ignore_javadoc`  
or `g:java_ignore_markdown` is defined.)

Do native shells on Windows support similar syntax?

**Review by zzzyxwvut — APPROVED** [review]

**Comment by Konfekt** [comment]
> Remember that the triple /// ‘shebang’ line will be matched
as javaMarkdownComment unless either g:java_ignore_javadoc
or g:java_ignore_markdown is defined

I guess that's fine in practice as it works on Bash and Zsh, but not Posix compliant.

I'd guess one is out of luck on Microsoft Windows regarding POSIX compatibilty if MSYS2 was not meant? Then I suppose it should work again

**Comment by chrisbra** [comment]
thanks!

**Pull request closed without merging**